listener.go

```diff
@@ -160,3 +188,90 @@ func (l *listener) logStats() {
 		}
 	}
 }
+
+// teeConn allows users to read from a connection, but retains the data read in teeBuf.
+type teeConn struct {
+	conn        net.Conn
+	teeReader   io.Reader
+	read        func([]byte) (int, error)
+	close       func() error
+	readLock    sync.Mutex
+	stopTeeOnce sync.Once
+}
+
+func newTeeConn(conn net.Conn) *teeConn {
+	var (
+		buf        = new(bytes.Buffer)
+		connReader = io.TeeReader(conn, buf)
+		teeReader  = io.MultiReader(buf, conn)
+	)
+	return &teeConn{conn, teeReader, connReader.Read, conn.Close, sync.Mutex{}, sync.Once{}}
+}
+
+func (tc *teeConn) Read(b []byte) (n int, err error) {
+	tc.readLock.Lock()
```

Comment:
I don't love locking on every read (and I'm not 100% sure it's necessary).

Reply:
It's "necessary" because the smux session is running a concurrent read loop.  The Read method it's calling needs to be swapped out concurrently.  I say necessary in quotations because, as far as I can tell, nothing bad would happen if the "wrong" Read method was called. We'd buffer slightly more data than necessary, but that'll get thrown away quickly anyway.  However, the race detector doesn't like it and that is a problem as it impedes our ability to use the race detector in any project importing `cmux`.

I'm not sure this issue is avoidable without adding serious complexity.  I think we'd have to replicate some of smux's internal logic to decide on our own whether to proceed with the connection.

What I can do is atomically load the read method and use an atomic swap when we stop the tee process.  If you go down the rabbit hole, the standard library is [already doing a few atomic operations](https://golang.org/src/internal/poll/fd_mutex.go#L117) when reading from a TCP connection.  I don't think adding one more atomic load operation should have much impact.  What do you think?

Reply:
Hmm although that would involve a type assertion on every read call which is probably worse.

Reply:
I'll write some benchmarks!

Reply:
I wrote some simple benchmarks, attempting to isolate the differences in each method:

```
type customReader struct {
	read func([]byte) (int, error)
}

func (cr customReader) Read(b []byte) (n int, err error) {
	return cr.read(b)
}

var data = make([]byte, 100)

func init() {
	rand.Read(data)
}

func BenchmarkSimple(b *testing.B) {
	r := customReader{
		read: func(b []byte) (int, error) {
			n := copy(b, data)
			return n, nil
		},
	}
	buf := make([]byte, len(data))

	b.ResetTimer()
	for i := 0; i < b.N; i++ {
		r.Read(buf)
	}
}

func BenchmarkLocking(b *testing.B) {
	m := new(sync.Mutex)
	r := customReader{
		read: func(b []byte) (int, error) {
			m.Lock()
			defer m.Unlock()
			n := copy(b, data)
			return n, nil
		},
	}
	buf := make([]byte, len(data))

	b.ResetTimer()
	for i := 0; i < b.N; i++ {
		r.Read(buf)
	}
}

func BenchmarkLockingNoDefer(b *testing.B) {
	m := new(sync.Mutex)
	r := customReader{
		read: func(b []byte) (int, error) {
			m.Lock()
			n := copy(b, data)
			m.Unlock()
			return n, nil
		},
	}
	buf := make([]byte, len(data))

	b.ResetTimer()
	for i := 0; i < b.N; i++ {
		r.Read(buf)
	}
}

func BenchmarkAtomicLoad(b *testing.B) {
	v := atomic.Value{}
	v.Store(func(b []byte) (int, error) {
		n := copy(b, data)
		return n, nil
	})
	r := customReader{
		read: func(b []byte) (int, error) {
			return v.Load().(func(b []byte) (int, error))(b)
		},
	}
	buf := make([]byte, len(data))

	b.ResetTimer()
	for i := 0; i < b.N; i++ {
		r.Read(buf)
	}
}
```

 As you suggested, getting rid of the defer call has a big impact.  Also, it looks like atomic loads are almost as good as no synchronization at all, even with the type assertion.

```
> go test -run=^$ -bench=. -benchmem
goos: darwin
goarch: amd64
pkg: github.com/getlantern/cmux
BenchmarkSimple-12            	171086394	         7.61 ns/op	       0 B/op	       0 allocs/op
BenchmarkLocking-12           	29329096	        38.1 ns/op	       0 B/op	       0 allocs/op
BenchmarkLockingNoDefer-12    	69478320	        16.6 ns/op	       0 B/op	       0 allocs/op
BenchmarkAtomicLoad-12        	120930295	        10.7 ns/op	       0 B/op	       0 allocs/op
PASS
ok  	github.com/getlantern/cmux	7.479s
```

Using 1 byte payloads doesn't change things much:

```
> go test -run=^$ -bench=. -benchmem
goos: darwin
goarch: amd64
pkg: github.com/getlantern/cmux
BenchmarkSimple-12            	267295834	         4.19 ns/op	       0 B/op	       0 allocs/op
BenchmarkLocking-12           	32703547	        37.7 ns/op	       0 B/op	       0 allocs/op
BenchmarkLockingNoDefer-12    	80736678	        14.3 ns/op	       0 B/op	       0 allocs/op
BenchmarkAtomicLoad-12        	169301498	         7.07 ns/op	       0 B/op	       0 allocs/op
PASS
ok  	github.com/getlantern/cmux	6.299s
```

Same goes for 1000 byte payloads:

```
> go test -run=^$ -bench=. -benchmem
goos: darwin
goarch: amd64
pkg: github.com/getlantern/cmux
BenchmarkSimple-12            	75809527	        17.6 ns/op	       0 B/op	       0 allocs/op
BenchmarkLocking-12           	23545557	        48.9 ns/op	       0 B/op	       0 allocs/op
BenchmarkLockingNoDefer-12    	47812003	        27.8 ns/op	       0 B/op	       0 allocs/op
BenchmarkAtomicLoad-12        	63111607	        19.4 ns/op	       0 B/op	       0 allocs/op
PASS
ok  	github.com/getlantern/cmux	5.281s
```

So I'll switch over to atomic loading and swapping.  I think it's simpler / cleaner anyway.

Reply:
Also, I tried to benchmark reading out of a TCP connection and attempted to isolate out the act of writing data into the connection.  It was a little tricky because you can only write so much before the write calls block (I'm guessing due to internal buffering?).  I was getting ~1000 ns/op.  Not sure how accurate that is since the benchmark was kind of complex.  Regardless, any of the synchronization implementations above would surely be low-impact compared to things like network latency.

Reply:
Okay made the change [here](https://github.com/getlantern/cmux/commit/3c32a5705d44a96b8c60a635d2dc18d8828b0a8b) and issued a PR to update http-proxy-lantern: https://github.com/getlantern/http-proxy-lantern/pull/347